File: src/main/java/io/github/seggan/jyxal/compiler/wrappers/MainMethod.java
package io.github.seggan.jyxal.compiler.wrappers;

import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.Label;

public class MainMethod extends JyxalMethod {

    private final Label start = new Label();
    private final Label end = new Label();

    MainMethod(ClassWriter cw, int access, String name, String desc) {
        super(cw, access, name, desc);

        stackVar = 1;
        ctxVar = 2;

        visitTypeInsn(NEW, "runtime/ProgramStack");
        visitInsn(DUP);
        visitVarInsn(ALOAD, 0);
        visitMethodInsn(INVOKESPECIAL, "runtime/ProgramStack", "<init>", "([Ljava/lang/String;)V", false);
        visitVarInsn(ASTORE, stackVar);

        mv.visitFieldInsn(
                GETSTATIC,
                "runtime/math/BigComplex",
                "ZERO",
                "Lruntime/math/BigComplex;"
        );
        visitVarInsn(ASTORE, ctxVar);

        visitLabel(start);
        visitLocalVariable("stack", "Lruntime/ProgramStack;", null, start, end, stackVar);

        visitLocalVariable("ctx", "Ljava/lang/Object;", null, start, end, ctxVar);
    }

    @Override
    public void visitEnd() {
        visitLabel(end);
        super.visitEnd();
    }
}

File: src/main/kotlin/io/github/seggan/jyxal/compiler/wrappers/MainMethod.kt
package io.github.seggan.jyxal.compiler.wrappers

import org.objectweb.asm.ClassWriter
import org.objectweb.asm.Label
import org.objectweb.asm.Opcodes

class MainMethod internal constructor(cw: ClassWriter, access: Int, name: String, desc: String) : JyxalMethod(cw, access, name, desc) {

    private val start = Label()
    private val end = Label()

    override fun visitEnd() {
        visitLabel(end)
        super.visitEnd()
    }

    init {
        stackVar = 1
        ctxVar = 2
        visitTypeInsn(Opcodes.NEW, "runtime/ProgramStack")
        visitInsn(Opcodes.DUP)
        visitVarInsn(Opcodes.ALOAD, 0)
        visitMethodInsn(Opcodes.INVOKESPECIAL, "runtime/ProgramStack", "<init>", "([Ljava/lang/String;)V", false)
        visitVarInsn(Opcodes.ASTORE, stackVar)
        visitFieldInsn(
            Opcodes.GETSTATIC,
            "runtime/math/BigComplex",
            "ZERO",
            "Lruntime/math/BigComplex;"
        )
        visitVarInsn(Opcodes.ASTORE, ctxVar)
        visitLabel(start)
        visitLocalVariable("stack", "Lio/github/seggan/jyxal/runtime/ProgramStack;", null, start, end, stackVar)
        visitLocalVariable("ctx", "Ljava/lang/Object;", null, start, end, ctxVar)
    }
}